extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -0,0 +1,357 @@
+     *    subject to race conditions.
+     *  - If a key promise (i.e., the value of a property access on the `keys`
+     *    object) is absent (i.e., `undefined`), then no key with that keypair
+     *    ID is present (but this could be because of a stale cache). If it's
+     *    null, then no key with that keypair ID could be found even after
+     *    cachebusting. If it's a `Promise` that resolves to `null`, then key
+     *    data for that keypair ID was found but could not be imported
+     *    successfully; this most likely indicates signing service misbehavior.
+     *    The success case is a `Promise` that resolves to a `CryptoKey`.
+     *
+     * @private {?Object<string, {promise: !Promise<boolean>, keys: !Object<string, ?Promise<?webCrypto.CryptoKey>>}>}
+     */
+    this.signers_ = cryptoFor(win).isPkcsAvailable() ? {} : null;
+  }
+
+  /**
+   * Fetches and imports the public keyset for the named signing service,
+   * without any cachebusting. Hopefully, this will hit cache in many cases
+   * and not make an actual network round-trip. This method should be called
+   * as early as possible, once it's known which signing service is likely to
+   * be used, so that the network request and key imports can execute in
+   * parallel with other operations.
+   *
+   * @param {string} signingServiceName
+   */
+  loadKeyset(signingServiceName) {
+    if (this.signers_ && !this.signers_[signingServiceName]) {
+      const keys = {};
+      const promise = this.fetchAndAddKeys_(keys, signingServiceName, null);
+      this.signers_[signingServiceName] = {promise, keys};
+    }
+  }
+
+  /**
+   * Verifies that `signature` is the correct cryptographic signature for
+   * `creative`, with the public key from the named signing service identified
+   * by `keypairId`.
+   *
+   * As a precondition, `loadKeyset(signingServiceName)` must have already been
+   * called.
+   *
+   * If the keyset for the named signing service was imported successfully but
+   * did not include a key for `keypairId`, this may be the result of a stale
+   * browser cache. To work around this, `keypairId` is added to the public key
+   * endpoint URL as a query parameter and the keyset is re-fetched. Other kinds
+   * of failures, including network connectivity failures, are not retried.
+   *
+   * @param {string} signingServiceName
+   * @param {string} keypairId
+   * @param {!Uint8Array} signature
+   * @param {!ArrayBuffer} creative
+   * @return {!Promise<?VerificationFailure>} resolves to `null` on success, or,
+   *     on failure, to a `VerificationFailure` indicating the cause
+   */
+  verify(signingServiceName, keypairId, signature, creative) {
+    if (this.signers_) {
+      const signer = this.signers_[signingServiceName];
+      return signer.promise.then(success => {
+        if (success) {
+          const keyPromise = signer.keys[keypairId];
+          if (keyPromise === undefined) {
```

Comment:
Do you mean changing the type of this value, or creating another parallel data structure? I'd have greatly preferred to use a tagged union instead of just a null comparison, but Closure doesn't support those and I _really_ don't want to litter this code with type assertions (the compile-time type checking is a lot of how I know it's correct at all).